File: jscomp/others/belt_MapDict.ml
(* Copyright (C) 2017 Authors of BuckleScript
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

module N = Belt_internalAVLtree
module A = Belt_Array

type ('key,  'a, 'id) t = ('key, 'a) N.t

type ('key, 'id) cmp = ('key, 'id)  Belt_Id.cmp

let empty = N.empty
let fromArray = N.fromArray
let isEmpty = N.isEmpty
let cmp = N.cmp
let cmpU = N.cmpU             
let eq = N.eq
let eqU = N.eqU  
let has = N.has

let forEach = N.forEach
let forEachU = N.forEachU
let reduce = N.reduce
let reduceU = N.reduceU
let every = N.every                
let everyU = N.everyU
let some = N.some 
let someU = N.someU
              
let size = N.size
let toList = N.toList
let toArray = N.toArray
let keysToArray = N.keysToArray
let valuesToArray = N.valuesToArray

let minimum = N.minimum
let maximum = N.maximum
let minKey = N.minKey
let maxKey = N.maxKey
let minKeyUndefined = N.minKeyUndefined
let maxKeyUndefined = N.maxKeyUndefined
let minUndefined = N.minUndefined
let maxUndefined = N.maxUndefined
let get = N.get
let getUndefined = N.getUndefined
let getWithDefault = N.getWithDefault
let getExn = N.getExn

let mapWithKey = N.mapWithKey
let mapWithKeyU = N.mapWithKeyU

let mapU  = N.mapU
let map = N.map
let keep = N.keepShared 
let keepU = N.keepSharedU
let partitionU = N.partitionSharedU
let partition = N.partitionShared                   
let checkInvariantInternal = N.checkInvariantInternal
let rec set  (t : _ t) newK newD  ~cmp =
  match N.toOpt t with 
  | None -> N.singleton newK newD 
  | Some n  ->
    let k= N.key n in 
    let c = (Belt_Id.getCmpInternal cmp) newK k [@bs] in
    if c = 0 then
      N.return (N.updateValue n newD) 
    else 
      let l,r,v = N.left n, N.right n, N.value n in 
      if c < 0 then (* Worth optimize for reference equality? *)
        N.bal (set ~cmp l newK newD ) k v  r
      else
        N.bal l k v (set ~cmp r newK newD )

let rec updateU  (t : _ t) newK f  ~cmp :  _ t =
  match N.toOpt t with 
  | None ->
    begin match f None [@bs] with 
      | None -> t 
      | Some newD -> N.singleton newK newD 
    end 
  | Some n  ->
    let k= N.key n in 
    let c = (Belt_Id.getCmpInternal cmp) newK k [@bs] in
    if c = 0 then
      match f (Some (N.value n)) [@bs] with 
      | None ->
        let l, r = N.left n , N.right n in  
        begin match N.toOpt l, N.toOpt r with
        | None, _ -> r
        | _, None -> l
        | _, Some rn ->
          let kr, vr = ref (N.key rn), ref (N.value rn) in
          let r = N.removeMinAuxWithRef rn kr vr in
          N.bal l !kr !vr r 
        end
      | Some newD -> N.return (N.updateValue n newD)
    else 
      let l,r,v = N.left n, N.right n, N.value n in 
      if c < 0 then
        let ll = (updateU ~cmp l newK f ) in
        if l == ll then
          t
        else 
          N.bal ll k v  r            
      else
        let rr = (updateU ~cmp r newK f) in
        if r == rr then t 
        else N.bal l k v rr

let update t newK f ~cmp =
  updateU t newK (fun [@bs] a  -> f a ) ~cmp
    
(*  unboxing API was not exported
    since the correct API is really awkard
    [bool -> 'k Js.null -> ('a Js.null * bool)]
    even for specialized [k] the first [bool] can 
    be erased, maybe the perf boost does not justify the inclusion of such API

    [updateWithNull m x f]
    the callback to [f exist v] 
    when [v] is non-null,
    [exist] is guaranteed to be true
    [v] is guranteed to be [null],
    when [exist] is [true], [v] could be [null], 
    since ['a] is polymorphic
*)


let rec removeAux0  n x ~cmp = 
  let l,v,r = N.(left n, key n, right n ) in 
  let c = (Belt_Id.getCmpInternal cmp) x v [@bs] in
  if c = 0 then
    match N.toOpt l, N.toOpt r with 
    | None, _ -> r 
    | _, None -> l 
    | _, Some rn -> 
      let kr, vr = ref (N.key rn), ref (N.value rn) in 
      let r = N.removeMinAuxWithRef rn kr vr in 
      N.bal l !kr !vr r
  else if c < 0 then
    match N.toOpt l with 
    | None -> N.return n (* Nothing to remove *)
    | Some left ->
      let ll = removeAux0 left x ~cmp in 
      if ll == l then (N.return n)
      else N.bal ll v (N.value n) r
  else
    match N.toOpt r with 
    | None -> N.return n (* Nothing to remove *)
    | Some right -> 
      let rr = removeAux0 ~cmp right x in
      if rr == r then N.return n
      else N.bal l v (N.value n)  rr


let remove n x ~cmp = 
  match N.toOpt n with        
  | None -> N.empty
  | Some n -> removeAux0 n x ~cmp 
  
let mergeMany   h arr ~cmp =   
  let len = A.length arr in 
  let v = ref h in  
  for i = 0 to len - 1 do 
    let key,value = A.getUnsafe arr i in 
    v := set !v  ~cmp key value
  done ;
  !v 

let rec splitAuxPivot n x pres  ~cmp =  
  let l,v,d,r = N.(left n , key n, value n, right n) in  
  let c = (Belt_Id.getCmpInternal cmp) x v [@bs] in 
  if c = 0 then begin 
    pres := Some d; 
    (l,  r)
  end
  else     
  if c < 0 then
    match N.toOpt l with 
    | None -> 
      N.(empty, return n)
    | Some l -> 
      let (ll,rl) = splitAuxPivot ~cmp l x pres in
      (ll,  N.join rl v d r)
  else
    match N.toOpt r with 
    | None ->
      N.(return n, empty)
    | Some r -> 
      let (lr,  rr) = splitAuxPivot ~cmp r x pres in
      (N.join l v d lr,  rr)


let split  n x ~cmp = 
  match N.toOpt n with 
  | None ->     
    N.(empty, empty), None
  | Some n  ->
    let pres = ref None in
    let v = splitAuxPivot ~cmp n x pres in 
    v, !pres

let rec mergeU s1 s2 f ~cmp =
  match N.(toOpt s1, toOpt s2) with
    (None, None) -> N.empty
  | Some _, None -> 
    N.keepMapU s1 (fun[@bs] k v -> 
        f k (Some v) None [@bs]
      )
  | None, Some _ -> 
    N.keepMapU s2 (fun[@bs] k v -> 
        f k None (Some v) [@bs]
      )
  | Some s1n , Some s2n -> 
    if N.height s1n  >= N.height s2n  then
      let l1, v1, d1, r1 = N.(left s1n, key s1n, value s1n, right s1n) in 
      let d2 = ref None in 
      let (l2, r2) = splitAuxPivot ~cmp s2n v1 d2 in
      let d2 = !d2 in 
      let newLeft = mergeU ~cmp l1 l2 f in 
      let newD = f v1 (Some d1) d2 [@bs] in 
      let newRight = mergeU ~cmp r1 r2 f in 
      N.concatOrJoin newLeft v1 newD  newRight
    else
      let l2,v2,d2,r2 = N.(left s2n, key s2n, value s2n, right s2n) in 
      let d1 = ref None in 
      let (l1,  r1) = splitAuxPivot ~cmp s1n v2 d1 in
      let d1 = !d1 in 
      let newLeft = mergeU ~cmp l1 l2 f in 
      let newD = (f v2 d1 (Some d2) [@bs]) in 
      let newRight = (mergeU ~cmp r1 r2 f) in 
      N.concatOrJoin newLeft v2 newD newRight

let merge s1 s2 f ~cmp =
  mergeU s1 s2 (fun[@bs] a b c -> f a b c) ~cmp
    
let rec removeMany0 t xs i len ~cmp =
  if i < len then
    let ele = A.getUnsafe xs i in
    let u =  removeAux0 t ele ~cmp in
    match N.toOpt u with
    | None -> u
    | Some t -> removeMany0 t xs (i + 1) len ~cmp 
  else
    N.return t
      
let removeMany t keys ~cmp =
  let len = A.length keys in
  match N.toOpt t with
  | None -> N.empty
  | Some t ->  removeMany0 t keys 0 len ~cmp 

File: lib/js/Belt_MapDict.js
'use strict';

let Primitive_option = require("./Primitive_option.js");
let Belt_internalAVLtree = require("./Belt_internalAVLtree.js");

function set(t, newK, newD, cmp) {
  if (t === undefined) {
    return Belt_internalAVLtree.singleton(newK, newD);
  }
  let k = t.k;
  let c = cmp(newK, k);
  if (c === 0) {
    return Belt_internalAVLtree.updateValue(t, newD);
  }
  let l = t.l;
  let r = t.r;
  let v = t.v;
  if (c < 0) {
    return Belt_internalAVLtree.bal(set(l, newK, newD, cmp), k, v, r);
  } else {
    return Belt_internalAVLtree.bal(l, k, v, set(r, newK, newD, cmp));
  }
}

function update(t, newK, f, cmp) {
  if (t !== undefined) {
    let k = t.k;
    let c = cmp(newK, k);
    if (c === 0) {
      let newD = f(Primitive_option.some(t.v));
      if (newD !== undefined) {
        return Belt_internalAVLtree.updateValue(t, Primitive_option.valFromOption(newD));
      }
      let l = t.l;
      let r = t.r;
      if (l === undefined) {
        return r;
      }
      if (r === undefined) {
        return l;
      }
      let kr = {
        contents: r.k
      };
      let vr = {
        contents: r.v
      };
      let r$1 = Belt_internalAVLtree.removeMinAuxWithRef(r, kr, vr);
      return Belt_internalAVLtree.bal(l, kr.contents, vr.contents, r$1);
    }
    let l$1 = t.l;
    let r$2 = t.r;
    let v = t.v;
    if (c < 0) {
      let ll = update(l$1, newK, f, cmp);
      if (l$1 === ll) {
        return t;
      } else {
        return Belt_internalAVLtree.bal(ll, k, v, r$2);
      }
    }
    let rr = update(r$2, newK, f, cmp);
    if (r$2 === rr) {
      return t;
    } else {
      return Belt_internalAVLtree.bal(l$1, k, v, rr);
    }
  }
  let newD$1 = f(undefined);
  if (newD$1 !== undefined) {
    return Belt_internalAVLtree.singleton(newK, Primitive_option.valFromOption(newD$1));
  } else {
    return t;
  }
}

function removeAux0(n, x, cmp) {
  let v = n.k;
  let l = n.l;
  let r = n.r;
  let c = cmp(x, v);
  if (c === 0) {
    if (l === undefined) {
      return r;
    }
    if (r === undefined) {
      return l;
    }
    let kr = {
      contents: r.k
    };
    let vr = {
      contents: r.v
    };
    let r$1 = Belt_internalAVLtree.removeMinAuxWithRef(r, kr, vr);
    return Belt_internalAVLtree.bal(l, kr.contents, vr.contents, r$1);
  }
  if (c < 0) {
    if (l === undefined) {
      return n;
    }
    let ll = removeAux0(l, x, cmp);
    if (ll === l) {
      return n;
    } else {
      return Belt_internalAVLtree.bal(ll, v, n.v, r);
    }
  }
  if (r === undefined) {
    return n;
  }
  let rr = removeAux0(r, x, cmp);
  if (rr === r) {
    return n;
  } else {
    return Belt_internalAVLtree.bal(l, v, n.v, rr);
  }
}

function remove(n, x, cmp) {
  if (n !== undefined) {
    return removeAux0(n, x, cmp);
  }
  
}

function mergeMany(h, arr, cmp) {
  let len = arr.length;
  let v = h;
  for (let i = 0; i < len; ++i) {
    let match = arr[i];
    v = set(v, match[0], match[1], cmp);
  }
  return v;
}

function splitAuxPivot(n, x, pres, cmp) {
  let v = n.k;
  let d = n.v;
  let l = n.l;
  let r = n.r;
  let c = cmp(x, v);
  if (c === 0) {
    pres.contents = Primitive_option.some(d);
    return [
      l,
      r
    ];
  }
  if (c < 0) {
    if (l === undefined) {
      return [
        undefined,
        n
      ];
    }
    let match = splitAuxPivot(l, x, pres, cmp);
    return [
      match[0],
      Belt_internalAVLtree.join(match[1], v, d, r)
    ];
  }
  if (r === undefined) {
    return [
      n,
      undefined
    ];
  }
  let match$1 = splitAuxPivot(r, x, pres, cmp);
  return [
    Belt_internalAVLtree.join(l, v, d, match$1[0]),
    match$1[1]
  ];
}

function split(n, x, cmp) {
  if (n === undefined) {
    return [
      [
        undefined,
        undefined
      ],
      undefined
    ];
  }
  let pres = {
    contents: undefined
  };
  let v = splitAuxPivot(n, x, pres, cmp);
  return [
    v,
    pres.contents
  ];
}

function merge(s1, s2, f, cmp) {
  if (s1 === undefined) {
    if (s2 !== undefined) {
      return Belt_internalAVLtree.keepMap(s2, (k, v) => f(k, undefined, Primitive_option.some(v)));
    } else {
      return;
    }
  }
  if (s2 === undefined) {
    return Belt_internalAVLtree.keepMap(s1, (k, v) => f(k, Primitive_option.some(v), undefined));
  }
  if (s1.h >= s2.h) {
    let v1 = s1.k;
    let d1 = s1.v;
    let l1 = s1.l;
    let r1 = s1.r;
    let d2 = {
      contents: undefined
    };
    let match = splitAuxPivot(s2, v1, d2, cmp);
    let d2$1 = d2.contents;
    let newLeft = merge(l1, match[0], f, cmp);
    let newD = f(v1, Primitive_option.some(d1), d2$1);
    let newRight = merge(r1, match[1], f, cmp);
    return Belt_internalAVLtree.concatOrJoin(newLeft, v1, newD, newRight);
  }
  let v2 = s2.k;
  let d2$2 = s2.v;
  let l2 = s2.l;
  let r2 = s2.r;
  let d1$1 = {
    contents: undefined
  };
  let match$1 = splitAuxPivot(s1, v2, d1$1, cmp);
  let d1$2 = d1$1.contents;
  let newLeft$1 = merge(match$1[0], l2, f, cmp);
  let newD$1 = f(v2, d1$2, Primitive_option.some(d2$2));
  let newRight$1 = merge(match$1[1], r2, f, cmp);
  return Belt_internalAVLtree.concatOrJoin(newLeft$1, v2, newD$1, newRight$1);
}

function removeMany(t, keys, cmp) {
  let len = keys.length;
  if (t !== undefined) {
    let _t = t;
    let _i = 0;
    while (true) {
      let i = _i;
      let t$1 = _t;
      if (i >= len) {
        return t$1;
      }
      let ele = keys[i];
      let u = removeAux0(t$1, ele, cmp);
      if (u === undefined) {
        return u;
      }
      _i = i + 1 | 0;
      _t = u;
      continue;
    };
  }
  
}

let empty;

let isEmpty = Belt_internalAVLtree.isEmpty;

let has = Belt_internalAVLtree.has;

let cmpU = Belt_internalAVLtree.cmp;

let cmp = Belt_internalAVLtree.cmp;

let eqU = Belt_internalAVLtree.eq;

let eq = Belt_internalAVLtree.eq;

let findFirstByU = Belt_internalAVLtree.findFirstBy;

let findFirstBy = Belt_internalAVLtree.findFirstBy;

let forEachU = Belt_internalAVLtree.forEach;

let forEach = Belt_internalAVLtree.forEach;

let reduceU = Belt_internalAVLtree.reduce;

let reduce = Belt_internalAVLtree.reduce;

let everyU = Belt_internalAVLtree.every;

let every = Belt_internalAVLtree.every;

let someU = Belt_internalAVLtree.some;

let some = Belt_internalAVLtree.some;

let size = Belt_internalAVLtree.size;

let toList = Belt_internalAVLtree.toList;

let toArray = Belt_internalAVLtree.toArray;

let fromArray = Belt_internalAVLtree.fromArray;

let keysToArray = Belt_internalAVLtree.keysToArray;

let valuesToArray = Belt_internalAVLtree.valuesToArray;

let minKey = Belt_internalAVLtree.minKey;

let minKeyUndefined = Belt_internalAVLtree.minKeyUndefined;

let maxKey = Belt_internalAVLtree.maxKey;

let maxKeyUndefined = Belt_internalAVLtree.maxKeyUndefined;

let minimum = Belt_internalAVLtree.minimum;

let minUndefined = Belt_internalAVLtree.minUndefined;

let maximum = Belt_internalAVLtree.maximum;

let maxUndefined = Belt_internalAVLtree.maxUndefined;

let get = Belt_internalAVLtree.get;

let getUndefined = Belt_internalAVLtree.getUndefined;

let getWithDefault = Belt_internalAVLtree.getWithDefault;

let getExn = Belt_internalAVLtree.getExn;

let checkInvariantInternal = Belt_internalAVLtree.checkInvariantInternal;

let updateU = update;

let mergeU = merge;

let keepU = Belt_internalAVLtree.keepShared;

let keep = Belt_internalAVLtree.keepShared;

let partitionU = Belt_internalAVLtree.partitionShared;

let partition = Belt_internalAVLtree.partitionShared;

let mapU = Belt_internalAVLtree.map;

let map = Belt_internalAVLtree.map;

let mapWithKeyU = Belt_internalAVLtree.mapWithKey;

let mapWithKey = Belt_internalAVLtree.mapWithKey;

exports.empty = empty;
exports.isEmpty = isEmpty;
exports.has = has;
exports.cmpU = cmpU;
exports.cmp = cmp;
exports.eqU = eqU;
exports.eq = eq;
exports.findFirstByU = findFirstByU;
exports.findFirstBy = findFirstBy;
exports.forEachU = forEachU;
exports.forEach = forEach;
exports.reduceU = reduceU;
exports.reduce = reduce;
exports.everyU = everyU;
exports.every = every;
exports.someU = someU;
exports.some = some;
exports.size = size;
exports.toList = toList;
exports.toArray = toArray;
exports.fromArray = fromArray;
exports.keysToArray = keysToArray;
exports.valuesToArray = valuesToArray;
exports.minKey = minKey;
exports.minKeyUndefined = minKeyUndefined;
exports.maxKey = maxKey;
exports.maxKeyUndefined = maxKeyUndefined;
exports.minimum = minimum;
exports.minUndefined = minUndefined;
exports.maximum = maximum;
exports.maxUndefined = maxUndefined;
exports.get = get;
exports.getUndefined = getUndefined;
exports.getWithDefault = getWithDefault;
exports.getExn = getExn;
exports.checkInvariantInternal = checkInvariantInternal;
exports.remove = remove;
exports.removeMany = removeMany;
exports.set = set;
exports.updateU = updateU;
exports.update = update;
exports.mergeU = mergeU;
exports.merge = merge;
exports.mergeMany = mergeMany;
exports.keepU = keepU;
exports.keep = keep;
exports.partitionU = partitionU;
exports.partition = partition;
exports.split = split;
exports.mapU = mapU;
exports.map = map;
exports.mapWithKeyU = mapWithKeyU;
exports.mapWithKey = mapWithKey;
/* No side effect */
